src/resolve/resolved-dns-query.c

```diff
@@ -480,6 +480,12 @@ int dns_query_new(
 
         assert(m);
 
+        /* Check for records that is refused and refuse query for the records if matched in configuration */
+        DNS_QUESTION_FOREACH(key, question_utf8)
+                if (set_contains(m->refuse_record_types, INT_TO_PTR(key->type))) {
+                        return log_debug_errno(SYNTHETIC_ERRNO(ENOANO), "Got request for %s record that is refused.", dns_type_to_string(key->type));
+                }
```

Comment:
coding style: no `{}` around single ine if blocks.

Hmm, so this is not going to work. The only reason why DnsQuestion structures contain multiple DnsResourceKey objects is to permit combined A + AAAA queries (because clients typically want to query both at once). However, if such a question is issued this loop of yours here, will refuse the whole query, even though instead we should simply suppress the relevant record type from the question, and if the result is non-empty continue.

Hence, maybe introduce a new helper call here, that "mangles" a DnsQuestion object: if any of the rr types is listed among the rr types to refuse it will remove them from the question, unless this would make the question empty in which case it would return an error.

Note that DnsResourceKey and DnsQuestion are supposed to be immutable once constructed fully. This means if you realize you need to remove a key from the question, then you need to allocate a new question object and initialize it from what's left.